src/services/GasEstimator.ts

```diff
@@ -0,0 +1,140 @@
+import { BlockWithTransactions, FeeData, TransactionResponse } from '@ethersproject/abstract-provider';
+import { BigNumber, ethers } from 'ethers';
+
+import { Type1GasEstimator } from './Type1GasEstimator';
+import { Type2GasEstimator } from './Type2GasEstimator';
+import { GasEstimation, EstimateParams } from '../types/GasEstimation';
+import { formatWeiToGwei } from '../utils/helpers';
+
+export class GasEstimator {
+  public static async apply(
+    provider: ethers.providers.Provider,
+    minGasPrice: number,
+    maxGasPrice: number
+  ): Promise<GasEstimation> {
+    const [block, currentGasPrice, feeData] = await GasEstimator.fetchProviderInfo(provider);
+
+    const [metrics, prices, maxPriorityFees] = GasEstimator.getGasMetricsFor(block);
+
+    return GasEstimator.estimate({
+      currentGasPrice: currentGasPrice.toNumber(),
+      feeData,
+      metrics,
+      prices,
+      minGasPrice,
+      maxGasPrice,
+      maxPriorityFees,
+    });
+  }
+
+  public static printable(metrics: GasEstimation): string {
+    const { gasPrice, maxPriorityFeePerGas, maxFeePerGas, min, max, avg, isTxType2, baseFeePerGas } = metrics;
+
+    return (
+      `isTxType2: ${isTxType2 ? 'yes' : 'no'}, ` +
+      `gasPrice: ${formatWeiToGwei(gasPrice)} Gwei, ` +
+      `baseFeePerGas: ${baseFeePerGas > 1e6 ? formatWeiToGwei(baseFeePerGas) + ' Gwei' : baseFeePerGas}, ` +
+      `maxPriorityFee: ${maxPriorityFeePerGas ? formatWeiToGwei(maxPriorityFeePerGas) : '-'} Gwei, ` +
+      `maxFee: ${maxFeePerGas ? formatWeiToGwei(maxFeePerGas) : '-'} Gwei, ` +
+      `min: ${formatWeiToGwei(min)} Gwei, ` +
+      `max: ${formatWeiToGwei(max)} Gwei, ` +
+      `avg: ${formatWeiToGwei(avg)} Gwei`
+    );
+  }
+
+  private static async fetchProviderInfo(
+    provider: ethers.providers.Provider
+  ): Promise<[BlockWithTransactions, BigNumber, FeeData]> {
+    return Promise.all([provider.getBlockWithTransactions('latest'), provider.getGasPrice(), provider.getFeeData()]);
+  }
+
+  private static getGasMetricsFor(
+    block: BlockWithTransactions
+  ): [gasPriceMetrics: GasEstimation, prices: number[], maxPriorityFees: number[]] {
+    const isTxType2 = Boolean(block.baseFeePerGas);
+    const transactions = block.transactions.filter(GasEstimator.containsGasPrice);
+    const prices = transactions.map(GasEstimator.getGasPrice);
+
+    if (!prices.length) {
```

Comment:
Does it happen because transactions are empty?

Reply:
Actually when there **are** prices. I'll leave it more explicit

Reply:
Done. I think it's clearer now